dev/io.openliberty.org.jboss.resteasy.common/src/io/openliberty/restfulWS/introspector/LibertyResteasyEndpointIntrospector.java

```diff
@@ -0,0 +1,93 @@
+
+import java.io.PrintWriter;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.ConfigurationPolicy;
+
+import com.ibm.ws.container.service.app.deploy.ApplicationInfo;
+import com.ibm.ws.container.service.state.ApplicationStateListener;
+import com.ibm.ws.container.service.state.StateChangeException;
+import com.ibm.wsspi.logging.Introspector;
+
+@Component(service = { ApplicationStateListener.class, Introspector.class, LibertyResteasyEndpointLoggingIntrospector.class },
+           immediate = true,
+           configurationPolicy = ConfigurationPolicy.IGNORE,
+           property = "service.vendor=IBM")
+public class LibertyResteasyEndpointIntrospector implements ApplicationStateListener, Introspector, LibertyResteasyEndpointLoggingIntrospector {
+
+    private static final Map<String, List<String>> applicationEndpoints = new ConcurrentHashMap<String, List<String>>();
+
+    @Override
+    public void applicationStarting(ApplicationInfo appInfo) throws StateChangeException {
+        // add the application to the map
+        applicationEndpoints.put(appInfo.getName(), new ArrayList<String>());
+    }
+
+    @Override
+    public void applicationStarted(ApplicationInfo appInfo) throws StateChangeException {}
+
+    @Override
+    public void applicationStopping(ApplicationInfo appInfo) {}
+
+    @Override
+    public void applicationStopped(ApplicationInfo appInfo) {
+        // remove all endpoints for the application to prevent a memory leak
+        applicationEndpoints.remove(appInfo.getName());
+    }
+
+    @Override
+    public String getIntrospectorName() {
+        return "LibertyResteasyEndpointIntrospector";
+    }
+
+    @Override
+    public String getIntrospectorDescription() {
+        return "Log the JAX-RS endpoints for each Application.";
+    }
+
+    @Override
+    public void introspect(PrintWriter out) throws Exception {
+        for (String key : applicationEndpoints.keySet()) {
+            out.println(key + " endpoints:");
+            for (String ep : applicationEndpoints.get(key)) {
+                out.println(ep);
+            }
+            out.println();
+        }
+    }
+
+    @Override
+    public String addEndpoint(String appName, String httpMethod, String endpoint, String className, String methodName) {
+        StringBuilder sb = new StringBuilder();
+        sb.append("- ");
+        sb.append(httpMethod);
+        sb.append(" ");
+        sb.append(endpoint);
+        sb.append(" - ");
+        sb.append(className);
+        sb.append("#");
+        sb.append(methodName);
+        sb.append("()");
+        String endpointInfo = sb.toString();
+        
+        // appName key and it's List should always exist because we initialized during applicationStarting
+        applicationEndpoints.get(appName).add(endpointInfo);
```

Comment:
I don't think we can assume this.  This is probably something that we would want to trace in the event that the app name didn't match the key set of known apps.  So maybe something like this:

```suggestion
        // appName key and it's List should always exist because we initialized during applicationStarting
        List<String> endpoints = applicationEndpoints.get(appName);
        if (endpoints != null) {
            endpoints.add(endpointInfo);
        } else if (TraceComponent.isAnyTraceEnabled() && tc.isDebugEnabled()) {
            Tr.debug(tc, "Attempting to add endpoint to unknown application, " + appName + ". Known apps: " + applicationEndpoints.keySet());
        }
```

You'll need to add the `tc` stuff for this to work - but it avoids any memory leak and allows us to figure out why an app might have a different name - or if we have some sort of bug where we attempt to add an endpoint before an app has started or after it has been shut down.